### DISCUSSION TITLE
StreamingResponse with file-like objects CPU-bound and slow

### DISCUSSION THREAD

**Discussion by lesinigo** [fastapi/fastapi#11932] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#!/usr/bin/env python3

import typing as t

from fastapi import FastAPI
from fastapi.responses import StreamingResponse
import uvicorn

app = FastAPI()


@app.get("/a")
async def serve_async() -> StreamingResponse:
    """test of "Using StreamingResponse with file-like objects" in an async method
    (this has been tested to be slow and CPU-bound)
    """
    the_file = open("100MiB.raw", "rb")

    def iterfile() -> t.Generator[bytes, None, None]:
        yield from the_file

    return StreamingResponse(iterfile())


@app.get("/b")
def serve_sync() -> StreamingResponse:
    """test of "Using StreamingResponse with file-like objects" in a non-async method
    (this has been tested to be slow and CPU-bound)
    """
    the_file = open("100MiB.raw", "rb")

    def iterfile() -> t.Generator[bytes, None, None]:
        yield from the_file

    return StreamingResponse(iterfile())


@app.get("/file/{filename}")
async def get_object(filename: str, file_storage: FileStorage = Depends(get_file_storage)) -> StreamingResponse:
    """actual implementation supporting both MinIO (via http) and local files"""

    the_file = file_storage.get(filename)

    if isinstance(the_file, urllib3.response.HTTPResponse):
        # the_file is an open http connection to MinIO
        # (this has been tested and is fast and with low CPU usage)
        content = the_file.stream(4 * 1024 * 1024)
        return StreamingResponse(content)

    # the_file is a standard python open(filename, "rb")
    # implementation is from the documentation but in an async method
    # https://fastapi.tiangolo.com/advanced/custom-response/#using-streamingresponse-with-file-like-objects
    # (this has been tested to be slow and CPU-bound)
    def iterfile() -> t.Generator[bytes, None, None]:
        yield from the_file

    return StreamingResponse(iterfile())


uvicorn.run(app)
```

### Description

I have a project based on FastAPI, running under the latest stable uvicorn version, deployed in multiple production environments using either Traefik v2, AWS ALB (on AWS EKS) and also tested with nothing in front of it (local tests/development on our workstations). I added an objectstorage-like feature that in production is using MinIO as backend and in local testing is using plain files in a local folder.

The same path operator is thus used to serve files whether they are from the MinIO-backed storage or from the local filesystem storage. Since MinIO is serving the file on an HTTP request we must use `StreamingResponse` and cannot use `FileResponse` and since the path operator is the same in both cases we are obliged to use `StreamingResponse` also when serving files from the local filesystem.

We copied the same code that's in the docs for [Using StreamingResponse with file-like objects](https://fastapi.tiangolo.com/advanced/custom-response/#using-streamingresponse-with-file-like-objects) as you can see in the example code quoted here, and we noted that when serving the response from an HTTP request it is really fast and uses little to no CPU, but when serving the response in the way of "Using StreamingResponse with file-like objects" it is really slow and cpu-bound (the python process goes to 100% of one cpu core).

For example on my Core i7 2.3GHz MacBookPro, running on localhost in plain HTTP (no encryption) and without any reverse proxy in front of uvicorn, I get ~2MiB/sec of download speed with 100% python CPU usage.

I'm failing to understand if I'm missing something or doing something wrong? How could I efficiently serve files from the local filesystem while still using `StreamingResponse` to have the same path operator compatible with my other storage backend that works over HTTP requests?

### Operating System

Linux, macOS

### Operating System Details

Python 3.11.x under Fedora 40, Alpine Linux 3.11, macOS 14.5

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

Python 3.11.9

### Additional Context

uvicorn v0.30.3

**Comment by MrTONYCHAN** [comment]
The small implicit chunks are the likely cause. Iterating a binary file with `yield from the_file` iterates by lines, not by a fixed multi-megabyte block. For arbitrary binary data this can create inefficient chunking and high Python overhead.

Use explicit chunks and close the file deterministically:

```python
def iter_file(path: str, chunk_size: int = 1024 * 1024):
    with open(path, "rb") as f:
        while chunk := f.read(chunk_size):
            yield chunk

@app.get("/file/{filename}")
def get_file(filename: str):
    return StreamingResponse(
        iter_file(filename),
        media_type="application/octet-stream",
    )
```

A synchronous generator is appropriate here: Starlette runs sync iteration in its thread pool, so the blocking file reads do not block the event loop.

If the local backend does not strictly need the same response class, `FileResponse` is preferable because it handles file metadata, range requests, and efficient file serving. The path operation can still select `FileResponse` for local files and `StreamingResponse` for the MinIO stream; both are valid `Response` objects and do not need to be the same concrete class.

Also benchmark outside macOS development reload/debug mode and vary the chunk size (for example 64 KiB–1 MiB). If performance remains critical, let nginx/object storage serve the bytes directly.

**Comment by Anshumaan-25** [comment]
One thing that stands out is the way the local file is being streamed:

"def iterfile():
    yield from the_file"

Compared to your MinIO implementation, the two sources aren't actually streaming data the same way. The MinIO response explicitly reads and yields large chunks (4 * 1024 * 1024 bytes), whereas the local file relies on the file object's default iterator.
I'd try making the local implementation use explicit chunked reads as well:

"def iterfile():
    while chunk := the_file.read(4 * 1024 * 1024):
        yield chunk"

This keeps the behavior consistent across both storage backends and significantly reduces the number of iterations and Python-level overhead involved in producing the response.
Another thing to keep in mind is that StreamingResponse is a generic streaming abstraction, while FileResponse is optimized specifically for local files. FileResponse can take advantage of server and OS-level optimizations that aren't available when data is yielded from a Python generator, so it's expected to be more efficient for files stored on disk.
If the explicit chunked reads improve throughput, that would suggest the bottleneck was primarily the iterator implementation rather than FastAPI itself. If performance remains roughly the same, I'd then investigate the interaction between StreamingResponse and the ASGI server (Uvicorn/Starlette), as the overhead may be occurring further down the response pipeline.

**Comment by lesinigo** [comment]
Yes we actually did fix the problem using bigger chunks for the local-file use case.

I do (and did) know that the correct type would be `FileResponse` but since I had a single path operator for both local files and MinIO I had to choose the common denominator and go with `StreamingResponse`.

I guess the problem was more about the example in FastAPI documentation (we copied it verbatim, at first) but as far as I can tell that specific example is now gone from the documentation so I am going to close this issue.
